File: components/credentials/src/main/java/org/cloudfoundry/credhub/requests/StringGenerationParameters.java
package org.cloudfoundry.credhub.requests;

import java.util.Objects;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonProperty.Access;
import org.cloudfoundry.credhub.ErrorMessages;
import org.cloudfoundry.credhub.exceptions.ParameterizedValidationException;
import org.cloudfoundry.credhub.generators.PassayStringCredentialGenerator;

import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_DEFAULT;

@JsonInclude(NON_DEFAULT)
public class StringGenerationParameters extends GenerationParameters {

  // Value Parameters
  @JsonProperty(access = Access.WRITE_ONLY)
  private Integer length = PassayStringCredentialGenerator.DEFAULT_LENGTH;

  private String username;

  private boolean excludeLower;
  private boolean excludeNumber;
  private boolean excludeUpper;
  private boolean includeSpecial;

  public int getLength() {
    return length;
  }

  public void setLength(final int length) {
    this.length = length;
  }

  public String getUsername() {
    return username;
  }

  public void setUsername(final String username) {
    this.username = username;
  }

  public boolean isIncludeSpecial() {
    return includeSpecial;
  }

  public void setIncludeSpecial(final boolean includeSpecial) {
    this.includeSpecial = includeSpecial;
  }

  public boolean isExcludeNumber() {
    return excludeNumber;
  }

  public void setExcludeNumber(final boolean excludeNumber) {
    this.excludeNumber = excludeNumber;
  }

  public boolean isExcludeUpper() {
    return excludeUpper;
  }

  public void setExcludeUpper(final boolean excludeUpper) {
    this.excludeUpper = excludeUpper;
  }

  public boolean isExcludeLower() {
    return excludeLower;
  }

  public void setExcludeLower(final boolean excludeLower) {
    this.excludeLower = excludeLower;
  }

  @JsonIgnore
  public boolean isValid() {
    return !(!includeSpecial
      && excludeNumber
      && excludeUpper
      && excludeLower
    );
  }

  @Override
  public void validate() {
    if (!isValid()) {
      throw new ParameterizedValidationException(ErrorMessages.EXCLUDES_ALL_CHARSETS);
    }
  }

  @Override
  public boolean equals(final Object o) {
    if (this == o) {
      return true;
    }

    if (o == null || getClass() != o.getClass()) {
      return false;
    }

    final StringGenerationParameters that = (StringGenerationParameters) o;
    return excludeLower == that.excludeLower &&
      excludeNumber == that.excludeNumber &&
      excludeUpper == that.excludeUpper &&
      includeSpecial == that.includeSpecial &&
      Objects.equals(length, that.length) &&
      Objects.equals(username, that.username);
  }

  @Override
  public int hashCode() {
    return Objects.hash(length, username, excludeLower, excludeNumber, excludeUpper, includeSpecial);
  }
}

File: components/credentials/src/main/kotlin/org/cloudfoundry/credhub/requests/StringGenerationParameters.kt
package org.cloudfoundry.credhub.requests

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonInclude.Include.NON_DEFAULT
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonProperty.Access
import org.cloudfoundry.credhub.ErrorMessages
import org.cloudfoundry.credhub.exceptions.ParameterizedValidationException
import org.cloudfoundry.credhub.generators.PassayStringCredentialGenerator
import java.util.Objects

@JsonInclude(NON_DEFAULT)
class StringGenerationParameters : GenerationParameters() {

    // Value Parameters
    @JsonProperty(access = Access.WRITE_ONLY)
    var length: Int = PassayStringCredentialGenerator.DEFAULT_LENGTH

    var username: String? = null

    var excludeLower: Boolean = false
    var excludeNumber: Boolean = false
    var excludeUpper: Boolean = false
    var includeSpecial: Boolean = false

    @JsonIgnore
    fun isValid(): Boolean {
        return !(
            !includeSpecial &&
                excludeNumber &&
                excludeUpper &&
                excludeLower
            )
    }

    fun isExcludeLower(): Boolean {
        return excludeLower
    }

    fun isExcludeNumber(): Boolean {
        return excludeNumber
    }

    fun isExcludeUpper(): Boolean {
        return excludeUpper
    }

    fun isIncludeSpecial(): Boolean {
        return includeSpecial
    }

    override fun validate() {
        if (!isValid()) {
            throw ParameterizedValidationException(ErrorMessages.EXCLUDES_ALL_CHARSETS)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        if (other == null || javaClass != other.javaClass) {
            return false
        }

        val that = other as StringGenerationParameters?
        return excludeLower == that!!.excludeLower &&
            excludeNumber == that.excludeNumber &&
            excludeUpper == that.excludeUpper &&
            includeSpecial == that.includeSpecial &&
            length == that.length &&
            username == that.username
    }

    override fun hashCode(): Int {
        return Objects.hash(length, username, excludeLower, excludeNumber, excludeUpper, includeSpecial)
    }
}
